CentralHub.WebUI/Pages/Rooms.razor

```diff
@@ -1,96 +1,587 @@
+    private string _trackerDescription = "Your tracker description"; 
+    private string? _selectedTrackerMacAddress = null;
+    
+    private async Task AddRoomModal()
     {
-        _rooms = (await RoomService.GetRoomsAsync(_cancellationTokenSource.Token)).ToList();
+        await ShowModal("addRoomModal");
+        StateHasChanged();
+    }
 
-        var dictionary = new Dictionary<Room, int>();
+    private async Task ConfirmAddRoom(String name, String description)
+    {
+        if(string.IsNullOrWhiteSpace(name))
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Please enter a name for the room");
+            return;
+        }
         
-        // FIXME: This is inefficient.
-        foreach (var room in _rooms)
+        await RoomService.AddRoomAsync(name!, description, _cancellationTokenSource.Token);
+        await HideModal("addRoomModal");
+        NavigationManager.NavigateTo(NavigationManager.Uri, forceLoad: true);
+    }
+
+    private async Task EditRoomModal(RoomWrapper wRoom)
+    {
+        if(wRoom == null)
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Room is null");
+            return;
+        }
+
+        _selectedWRoom = wRoom;
+        _roomName = _selectedWRoom!.Room.Name;
+        _roomDescription = _selectedWRoom.Room.Description;
+        wRoom!.TrackerCount = await GetTrackerCountAsync(wRoom.Room);
+        await ShowModal("editRoomModal");
+        StateHasChanged();
+    }
+
+    private async Task DeleteRoomModal(RoomWrapper wRoom)
+    {
+        if(wRoom == null)
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Room is null");
+            return;
+        }
+
+        await DeleteRoom(wRoom!);
+        await HideModal("editRoomModal");
+        StateHasChanged();
+    }
+
+    private async Task ConfirmEditRoom(RoomWrapper wRoom, String name, String description)
+    {
+        if(string.IsNullOrWhiteSpace(name))
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Please enter a name for the room");
+            return;
+        } 
+        if(wRoom == null)
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Room is null");
+            return;
+        }
+
+        await RoomService.UpdateRoomAsync(wRoom!.Room, name!, description, _cancellationTokenSource.Token);
+        await HideModal("editRoomModal");
+        NavigationManager.NavigateTo(NavigationManager.Uri, forceLoad: true);
+    }
+
+    private async Task AddTrackerModal(RoomWrapper wRoom)
+    {
+        if(wRoom == null)
+        {
+            await JSRuntime.InvokeVoidAsync("alert", "Room is null");
+            return;
+        }
+
+        _selectedWRoom = wRoom;
+        _selectedRoomId = wRoom!.Room.RoomId;
+        _unregisteredTrackers = await TrackerService.GetUnregisteredTrackersAsync(_cancellationTokenSource.Token);
+        _selectedTrackerMacAddress = _unregisteredTrackers[0].WifiMacAddress;
```

Comment:
This brokey, it throws an exception when there arent any unregistered trackers

Reply:
fixed